src/internal/request-batcher.ts

```diff
@@ -0,0 +1,126 @@
+import { ConnectionInfo, FetchJsonResponse } from '@ethersproject/web';
+
+import { DEFAULT_MAX_REQUEST_BATCH_SIZE } from '../util/const';
+import { JsonRpcRequest } from './internal-types';
+
+/**
+ * Internal class to enqueue requests and automatically send/process batches.
+ *
+ * The underlying batching mechanism is loosely based on ethers.js's
+ * `JsonRpcBatchProvider`.
+ *
+ * @internal
+ */
+export class RequestBatcher {
+  /** Timeout timer that periodically sends the pending batch. */
+  private pendingBatchTimer: NodeJS.Timer | undefined;
+
+  /**
+   * Array of enqueued requests along with the constructed promise handlers for
+   * each request.
+   */
+  private pendingBatch: Array<BatchRequest> | undefined;
+
+  constructor(
+    private readonly sendBatchFn: SendBatchFn,
+    private readonly connection: ConnectionInfo,
+    private readonly maxBatchSize = DEFAULT_MAX_REQUEST_BATCH_SIZE
+  ) {}
+
+  /**
+   * Enqueues the provided request. The batch is immediately sent if the maximum
+   * batch size is reached. Otherwise, the request is enqueued onto a batch that
+   * is sent after 10ms.
+   *
+   * Returns a promise that resolves with the result of the request.
+   */
+  async enqueueRequest(request: JsonRpcRequest): Promise<any> {
+    if (this.pendingBatch === undefined) {
+      this.pendingBatch = [];
+    }
```

Comment:
(Minor) The logic might be simpler if you just have `pendingBatch` be an empty array instead of `undefined` in the case where it's empty. Everything would work out the same (with you checking for 0 length instead of undefined), except that you wouldn't need this one if-block here.